src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -752,41 +830,32 @@ class NBDWatchCtx : public librbd::UpdateWatchCtx
     , io_ctx(_io_ctx)
     , image(_image)
     , size(_size)
-  { }
+    , notify_thread(*this, &NBDWatchCtx::notify_entry)
+  {
+    started = true;
+    notify_thread.create("rbd_notify");
+  }
 
-  ~NBDWatchCtx() override {}
+  ~NBDWatchCtx() override
+  {
+    if (started) {
+      dout(10) << __func__ << ": terminating" << dendl;
+      std::unique_lock locker{lock};
+      terminated = true;
+      cond.notify_all();
```

Comment:
@idryomov here are the changes
(looks like I lost the last week locally saved version, just composed it fresh today)

I expect that this cond.notify_all() to unblock the wait in wait_notify(). Unfortunately its not happening. I might be missing something here. Could you please help?

Reply:
> I expect that this cond.notify_all() to unblock the wait in wait_notify(). Unfortunately its not happening.

I think it's actually happening but because you are holding the lock across `join()`, it locks up with `wait_notify()`.

Reply:
Ah! good catch. Thanks!